cilium/CFP-30572-non-default-deny-policies.md

```diff
@@ -0,0 +1,175 @@
+# CFP-30572: Non-default-Deny Network Policy
+
+**SIG: SIG-Policy**
+
+**Begin Design Discussion:** 2024-02-02
+
+**Cilium Release:** 1.16
+
+**Authors:** Casey Callendrello <cdc@isovalent.com>
+
+## Summary
+
+Allow users to create network policies that do not drop non-matching traffic. 
+
+## Motivation
+
+Consider a large, multitenant cluster. The cluster administrator is distinct from
+the workload owners, and the workloads are diverse and existing.
+
+The administrator may wish to create a Cilium Clusterwide Network Policy (CCNP)
+that affects most or all existing workloads. For example, they may wish to
+- Globally grant access to a security scanner
+- Ensure that workloads can access kube-dns
+- Proxy all DNS requests for monitoring purposes
+- Deny access to sensitive IPs, such as cloud provider metadata APIs
+
+All of these can already be accomplished by the existing policy language.
+
+By default, all Cilium endpoints have unrestricted network access. If at least one
+network policy applies to an endpoint, then all traffic is denied except that
+explicitly allowed by policy.
+
+This presents a challenge for the cluster administrator. They cannot assume that
+existing, heterogeneous worloads already have network policies provisioned. So, by
+creating an administrative network policy, they may select endpoints that **previously
+were in an allow-all state**. Thus, they may unintentionally block legitimate
+traffic for an existing workload, causing an outage.
+
+## Goals
+
+* Ability to safely roll-out broad-based network policies, without the risk
+  of disrupting existing traffic.
+
+## Non-Goals
+
+* An iterative, iptables-style firewall rules engine
+* Arbitrary, multi-level rule priority
+
+## Proposal
+
+### Overview
+
+We propose a new field in the policy spec, `DefaultAction`. This field specifies
+whether or not an implicit default-deny rule should be created _on behalf of this policy_.
+
+This field is optional. The default value is `Deny`. This matches the existing behavior, wherein any network policy implicitly blocks all other traffic.
+
+The value `Pass` indicates that this policy should not create a default-deny rule.
```

Comment:
One thing I like about the idea of a string for this property is the possibility for us to reuse this same field in future for a per-policy / per-namespace audit mode feature. May be worth considering whether we want to leave that option open here with a new enum value for the field or whether we would expect such a selective audit mode feature to be implemented separately.

Reply:
From discussion out-of-band with @nathanjsweet , @learnitall , @squeed we acknowledged the potential overlap for describing audit mode in this way, but also identified that there are nuances in the way that audit mode applies which raise some uncertainty about whether embedding audit mode into individual policies is a good idea. Specifically, audit mode is typically applying on a per-endpoint basis rather than selecting specific traffic to apply to. While it's plausible for us to express this concept in the policy in this area, we don't want to tie the overall proposal here to the possibility of audit mode, so it makes sense to put this idea back on the shelf for now. If and when we have stronger motivation to address a more dynamic audit mode feature for individual endpoints/namespaces or even policies, we can revisit the solution space specifically for that feature.